From: Ayhan Aydin <nsd.project.dev@gmail.com>
Date: Sat, 25 Jul 2026 20:00:00 +0300
Subject: [PATCH 1/4] mm/filemap: Add NSD prefetch hook point

Add a single function call in filemap_read() to notify the NSD
prefetcher of I/O events. This allows NSD to learn access patterns
and issue proactive vfs_fadvise(WILLNEED) calls.

The hook is protected by CONFIG_NSD and compiles to nothing when NSD
is disabled. It passes only file, offset and length - no new locking
or performance impact on the hot path.

Signed-off-by: Ayhan Aydin <nsd.project.dev@gmail.com>
---
 mm/filemap.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/mm/filemap.c b/mm/filemap.c
index 12345678..1234abcd 100644
--- a/mm/filemap.c
+++ b/mm/filemap.c
@@ -36,6 +36,7 @@
 #include <linux/pipe_fs_i.h>
 #include <linux/splice.h>
 #include <linux/rcupdate_wait.h>
+#include <linux/nsd.h>
 #include <asm/pgalloc.h>
 #include <asm/tlb.h>
 @@ -2641,6 +2642,10 @@ ssize_t filemap_read(struct kiocb *iocb, struct iov_iter *iter,
 	if (!count)
 		goto out;
 
+	if (IS_ENABLED(CONFIG_NSD))
+		nsd_notify_read(iocb->ki_filp, iocb->ki_pos,
+				iov_iter_count(iter));
+
 	already_read = 0;
 	if (!iov_iter_count(iter))
 		goto out;
-- 
2.43.0
